File: HNQQPopMenu/HNQQPopmenu/HNQQPopMenuView.m
//
//  HNQQDropMenuView.m
//  HNQQPopMenu
//
//  Created by 许浩男 on 16/4/2.
//  Copyright © 2016年 Zakariyya. All rights reserved.
//

#import "HNQQPopMenuView.h"
#import "HNPopMenuTableViewCell.h"
#import "HNQQPopMenuManager.h"
#import "HNQQPopMenuModel.h"

static const CGFloat triangleHeight = 10.0f;
static const CGFloat margin = 10.0f;

@interface HNQQPopMenuView()<UITableViewDataSource,UITableViewDelegate>

@property (nonatomic,strong) NSArray *dataArr;

@property (nonatomic,assign) CGPoint startPoint;

@end

@implementation HNQQPopMenuView

- (instancetype)initWithView:(UIView *)view items:(NSArray *)itemArr{
    if (self = [super init]) {
        self.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.15];
        _menuWidth = 150.0f;
        _cellHeight = 40.0f;
        CGPoint centerPoint = view.center;
        CGFloat tableViewY = CGRectGetMaxY(view.frame) + margin - 0.5 * _cellHeight * itemArr.count;
        CGFloat tableViewX = centerPoint.x - _menuWidth * 0.5 + _menuWidth * 0.5;
        if (tableViewX < margin + _menuWidth * 0.5) {
            tableViewX = margin - _menuWidth * 0.5;
        }
        else if (tableViewX + _menuWidth > [UIScreen mainScreen].bounds.size.width - margin){
            tableViewX = [UIScreen mainScreen].bounds.size.width - margin - _menuWidth * 0.5;
        }
        _startPoint = CGPointMake(centerPoint.x, tableViewY - triangleHeight + 0.5 * _cellHeight * itemArr.count);
        _tableView = [[UITableView alloc] initWithFrame:CGRectMake(tableViewX, tableViewY, _menuWidth, _cellHeight * itemArr.count) style:UITableViewStylePlain];
        _tableView.backgroundColor = [UIColor whiteColor];
        _tableView.layer.cornerRadius = 10.0f;
        _tableView.layer.masksToBounds = YES;
        _tableView.dataSource = self;
        _tableView.delegate = self;
        _tableView.bounces = NO;
        _tableView.layer.anchorPoint = centerPoint.x > [UIScreen mainScreen].bounds.size.width * 0.5 ? CGPointMake(1.0f, 0.0f) :CGPointMake(0.0f, 0.0f);
        _tableView.transform = CGAffineTransformMakeScale(0.001, 0.001);
        _tableView.rowHeight = _cellHeight;
        _dataArr = itemArr;
        [self addSubview:_tableView];
        [self drawTriangleLayer];
    }
    return self;
}

- (void)drawTriangleLayer{
    CGFloat triangleLength = triangleHeight * 2.0 / 1.732;
    UIBezierPath *path = [UIBezierPath bezierPath];
    [path moveToPoint:_startPoint];
    [path addLineToPoint:CGPointMake(_startPoint.x - triangleLength * 0.5, _startPoint.y + triangleHeight)];
    [path addLineToPoint:CGPointMake(_startPoint.x + triangleLength * 0.5, _startPoint.y + triangleHeight)];
    CAShapeLayer *triangleLayer = [CAShapeLayer layer];
    triangleLayer.path = path.CGPath;
    triangleLayer.fillColor = [UIColor whiteColor].CGColor;
    triangleLayer.strokeColor = [UIColor whiteColor].CGColor;
    self.triangleLayer = triangleLayer;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.15f * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self.layer addSublayer:triangleLayer];
    });
}

- (void)setMenuWidth:(CGFloat)menuWidth{
    _menuWidth = menuWidth;
    CGRect tableViewFrame = _tableView.frame;
    CGPoint tableViewCenter = _tableView.center;
    tableViewFrame.size.width = _menuWidth;
    tableViewFrame.origin.x = tableViewCenter.x - _menuWidth * 0.5;
    if (tableViewFrame.origin.x < margin + _menuWidth * 0.5) {
        tableViewFrame.origin.x = margin - _menuWidth * 0.5;
    }
    else if (tableViewFrame.origin.x > [UIScreen mainScreen].bounds.size.width - margin){
        tableViewFrame.origin.x = [UIScreen mainScreen].bounds.size.width - margin - _menuWidth * 0.5;
    }
    _tableView.frame = tableViewFrame;
}

- (void)setCellHeight:(CGFloat)cellHeight{
    _cellHeight = cellHeight;
    _tableView.rowHeight = _cellHeight;
}

//- (void)drawRect:(CGRect)rect {
//    CGContextRef context = UIGraphicsGetCurrentContext();
//    CGContextBeginPath(context);
//    CGFloat triangleLength = triangleHeight * 2.0 / 1.732;
//    CGContextMoveToPoint(context, _startPoint.x, _startPoint.y);
//    CGContextAddLineToPoint(context, _startPoint.x - triangleLength * 0.5, _startPoint.y + triangleHeight);
//    CGContextAddLineToPoint(context, _startPoint.x + triangleLength * 0.5, _startPoint.y + triangleHeight);
//    CGContextClosePath(context);
//    [[UIColor whiteColor] setFill];
//    [[UIColor whiteColor] setStroke];
//    CGContextDrawPath(context, kCGPathFillStroke);
//}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
    [HNQQPopMenuManager dismiss];
}

#pragma mark - UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.dataArr.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    HNPopMenuTableViewCell *cell = [[HNPopMenuTableViewCell alloc] initWithTableView:tableView];
    HNQQPopMenuModel *model = self.dataArr[indexPath.row];
    cell.imageView.image = [UIImage imageNamed:model.imageName];
    cell.textLabel.text = model.title;
    return cell;
}

#pragma mark - UITableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    if ([self.delegate respondsToSelector:@selector(QQPopMenuView:didSelectRow:)]) {
        [self.delegate QQPopMenuView:self didSelectRow:indexPath.row];
        [HNQQPopMenuManager dismiss];
    }
    if (self.clickAction) {
        self.clickAction(indexPath.row);
        [HNQQPopMenuManager dismiss];
    }
}


@end

File: HNQQPopMenu-swift/HNQQPopMenu-swift/HNQQPopMenu/HNQQPopMenuView.swift
//
//  HNQQPopMenuView.swift
//  HNQQPopMenu-swift
//
//  Created by zakariyyaSv on 16/4/22.
//  Copyright © 2016年 Zakariyya. All rights reserved.
//

import UIKit

protocol HNQQPopMenuViewDelegate : class {
    func HNMenuDidSelectRow(menuView:HNQQPopMenuView,row : NSInteger);
}

class HNQQPopMenuView: UIView,UITableViewDelegate,UITableViewDataSource{
    private let triangleHeight : CGFloat = 10.0;
    private let margin : CGFloat = 10.0;
    private var dataArr : [HNQQPopMenuModel]!;
    private var startPoint : CGPoint!;
    var triangleLayer : CAShapeLayer!;
    var action : actionClosure?
    var dismissed = true
    var tableView = UITableView();
    weak var delegate : HNQQPopMenuViewDelegate?;
    
    var menuWidth : CGFloat! = 150.0 {
        didSet {
            if menuWidth != oldValue {
                var tableViewFrame = tableView.frame
                let tableViewCenter = tableView.center
                tableViewFrame.size.width = menuWidth
                tableViewFrame.origin.x = tableViewCenter.x - menuWidth * 0.5
                if tableViewFrame.origin.x < margin + menuWidth * 0.5 {
                    tableViewFrame.origin.x = margin - menuWidth * 0.5
                }
                else if tableViewFrame.origin.x > UIScreen.mainScreen().bounds.size.width - margin{
                    tableViewFrame.origin.x = UIScreen.mainScreen().bounds.size.width - margin - menuWidth * 0.5
                }
                tableView.frame = tableViewFrame

            }
        }
    }
    var cellHeight : CGFloat! = 40.0 {
        didSet {
            if cellHeight != oldValue {
                tableView.rowHeight = cellHeight
            }
            
        }
    }
    init(view : UIView,itemArr:Array<HNQQPopMenuModel>){
        super.init(frame: CGRectZero);
        self.backgroundColor = UIColor.blackColor().colorWithAlphaComponent(0.15)
        let centerPoint = view.center;
        let tableViewY : CGFloat = CGRectGetMaxY(view.frame) + margin - 0.5 * cellHeight * (CGFloat)(itemArr.count)
        var tableViewX = centerPoint.x - menuWidth * 0.5 + menuWidth * 0.5
        if tableViewX < margin + menuWidth * 0.5 {
            tableViewX = margin - menuWidth * 0.5
        }
        else if tableViewX + menuWidth > UIScreen.mainScreen().bounds.size.width - margin{
            tableViewX = UIScreen.mainScreen().bounds.size.width - margin - menuWidth * 0.5
        }
        startPoint = CGPointMake(centerPoint.x, (CGFloat)(tableViewY) - (CGFloat)(triangleHeight) + 0.5 * (CGFloat)(cellHeight) * (CGFloat)(itemArr.count))
        tableView.frame = CGRectMake(tableViewX, tableViewY, menuWidth, cellHeight * (CGFloat)(itemArr.count))
        tableView.backgroundColor = UIColor.whiteColor()
        tableView.layer.cornerRadius = 10.0
        tableView.layer.masksToBounds = true
        tableView.bounces = false
        tableView.layer.anchorPoint = centerPoint.x > UIScreen.mainScreen().bounds.size.width * 0.5 ? CGPointMake(1.0, 0) : CGPointMake(0.0, 0.0)
        tableView.transform = CGAffineTransformMakeScale(0.001, 0.001)
        tableView.rowHeight = cellHeight
        tableView.dataSource = self
        tableView.delegate = self
        self.addSubview(tableView)
        dataArr = itemArr
        self.drawTriangleLayer()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func drawTriangleLayer() {
        let triangleLength = triangleHeight * 2.0 / 1.732
        let path = UIBezierPath()
        path.moveToPoint(startPoint)
        path.addLineToPoint(CGPointMake(startPoint.x - triangleLength * 0.5, startPoint.y + triangleHeight))
        path.addLineToPoint(CGPointMake(startPoint.x + triangleLength * 0.5, startPoint.y + triangleHeight))
        triangleLayer = CAShapeLayer()
        triangleLayer.path = path.CGPath
        triangleLayer.fillColor = UIColor.whiteColor().CGColor
        triangleLayer.strokeColor = UIColor.whiteColor().CGColor
        let time = dispatch_time(DISPATCH_TIME_NOW, (Int64)(150 * NSEC_PER_MSEC))
        dispatch_after(time, dispatch_get_main_queue(), {
            () -> Void in
            self.layer.addSublayer(self.triangleLayer)
        })
        
    }
    
    func tableViewBackgroundColor(color:UIColor){
        triangleLayer.removeFromSuperlayer()
        self.drawTriangleLayer()
    }
    
    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        HNQQPopMenuManager.dismiss()
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataArr.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell : HNPopMenuTableViewCell = HNPopMenuTableViewCell.initWithTableView(tableView)
        let model = dataArr[indexPath.row]
        cell.imageView?.image = UIImage.init(named: model.imageName as String)
        cell.textLabel?.text = model.title
        return cell
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        self.tableView(tableView, didDeselectRowAtIndexPath: indexPath)
        delegate?.HNMenuDidSelectRow(self, row: indexPath.row)
        action?(indexPath.row)
        if dismissed {
            HNQQPopMenuManager.dismiss()
        }
    }
    
    func tableView(tableView: UITableView, didDeselectRowAtIndexPath indexPath: NSIndexPath) {
        
    }

}
